Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1323 @@
+</details>
+
+But enough with theory, let's go to the practice and learn more concepts as we go through it.
+
+## Our first parser
+
+Below is a minimal definition of a grammar file that Happy can compile.
+In addition, we inserted an extra `empty` rule which we will delete later, as it's there only to make Happy happy.
+
+Make sure to put it in a new `Parser.y` file.
+
+```happy
+{
+{-# LANGUAGE DeriveFoldable #-}
+module Parser
+  ( parseMiniML
+  ) where
+
+import Data.ByteString.Lazy.Char8 (ByteString)
+import Data.Maybe (fromJust)
+import Data.Monoid (First (..))
+
+import qualified Lexer as L
+}
+
+%name parseMiniML
+%tokentype { L.RangedToken }
+%error { parseError }
+%monad { L.Alex } { >>= } { pure }
+%lexer { lexer } { L.RangedToken L.EOF _ }
+
+%%
+
+empty : {}  -- only to get the file compiling; we will remove this
+
+{
+parseError :: L.RangedToken -> L.Alex a
+parseError _ = do
+  (L.AlexPn _ line column, _, _, _) <- L.alexGetInput
+  L.alexError $ "Parse error at line " <> show line <> ", column " <> show column
+
+lexer :: (L.RangedToken -> L.Alex a) -> L.Alex a
+lexer = (=<< L.alexMonadScan)
+}
+```
+
+Happy has a similar structure to Alex, where we insert Haskell definitions at the top and the bottom of the file, with the parser definition in the middle.
+Likewise, anything between `{` and `}` is going to be inlined as well.
+
+Let's quickly visit each part of this file, from top to bottom.
+* The header contains the code that will generate the header of our Haskell file. Happy will generate the primary parsing function, `parseMiniML`, which we export.
+  * `DeriveFoldable` allows us to write `deriving (Foldable)`, which will be helpful later.
+  * We also import some extra things that will be useful in utilities later.
+* The following lines instruct Happy on how exactly to generate the parser. The things we should provide to it include:
+  * One or more names for our parser. The syntax should be `%name PARSER_NAME [PRODUCTION]`. If the `PRODUCTION` is not given, it uses the first non-terminal that appears in the file.
+  * What is the type of our tokens with `%tokentype`.
+  * Which function should it call if there is a parse error with `%error`.
+  * What is the monad that should be used and its canonical operations with `%monad` (we reuse `Alex` for simplicity).
+  * Which function lexes tokens and how to extract the EOF token with `%lexer`.
+* `%%`, after which we must specify the production rules for the grammar.
+* A trailer to include more Haskell definitions, such as auxiliary files and the AST that will be built by the parser. You can also define those in another module if you prefer.
+
+### Tokens
+
+Since Happy operates on tokens, we must instruct it on how to match each token.
+This is done using `%token`, which must be placed before the `%%` in the file.
+We list each token name that we've defined in our lexer and a Haskell pattern between braces indicating how to match it.
```

Comment:
I found the "that we've defined in our lexer" a bit confusing because it went me looking for Alex-looking stuff on the left side, while it is on the right side.